src/libos/src/net/socket/unix/stream/endpoint.rs

```diff
@@ -100,6 +103,38 @@ impl Inner {
         Ok(())
     }
 
+    pub fn poll(&self) -> IoEvents {
+        let mut events = IoEvents::empty();
+
+        /// get the status of peer
+        if self.reader.is_peer_shutdown() {
+            events |= IoEvents::RDHUP;
+            if self.writer.is_peer_shutdown() {
+                events |= IoEvents::HUP;
+            }
+        }
+
+        let reader_events = self.reader.poll();
+        let writer_events = self.writer.poll();
+        events |= (reader_events & IoEvents::IN) | (writer_events & IoEvents::OUT);
```

Comment:
What exactly is the relationship between the events of an endpoint and its underlying reader and writer? If some kind of conversion or filter should be done, then `EndpointObserver::on_event` should also reflect this. Since we have full control over the underlying `Channel`, it is possible for us to make the events on `Channel` consistent with the events  on `Endpoint`.